docs/developer/best-practices/navigation.asciidoc

```diff
@@ -0,0 +1,203 @@
+
+*Rendering a link to a different {kib} app on its own would also cause a full page reload:*
+
+[source,typescript jsx]
+----
+const myLink = () => <a href={core.http.basePath.prepend(`/dashboard/my-dashboard`)}> Go to Dashboard </a>; 
+----
+
+A workaround could be to handle a click, prevent browser navigation and use `core.application.navigateToApp` api:
+
+[source,typescript jsx]
+----
+const MySPALink = () => 
+    <a 
+        href={core.http.basePath.prepend(`/dashboard/my-dashboard`)} 
+        onClick={(e) => {
+            e.preventDefault();
+            core.application.navigateToApp('dashboard', {path: '/my-dashboard'}); 
+        }}> 
+        Go to Dashboard 
+    </a>; 
+----
+
+
+As it would be too much boilerplate to do this for each {kib} link in your app, there is a handy HOC that helps with it: 
+{kib-repo}tree/{branch}/src/plugins/kibana_react/public/app_links/redirect_app_link.tsx#L49[RedirectAppLinks].
+
+[source,typescript jsx]
+----
+const MyApp = () => 
+  <RedirectAppLinks application={core.application}>
+    {/*...*/}
+    {/* navigations using this link will happen in SPA friendly way */}
+        <a href={core.http.basePath.prepend(`/dashboard/my-dashboard`)}>Go to Dashboard</a>
+    {/*...*/}
+  </RedirectAppLinks>
+----
+
+
+[[deep-linking]]
+=== Deep-linking into {kib} apps
+
+**Consider a {kib} app URL a part of app's plugin contract.** +
+Try to avoid hardcoding other app's URL in your app's code:
+
+[source,typescript jsx]
+----
+const urlToDashboarWithComplexState = core.http.basePath.prepend(`/dashboards#/view/0289f5d0-e61c-11ea-8c31-8bd3114ed023?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-15m,to:now))&_a=(description:'',filters:!(),fullScreenMode:!f,options:(hidePanelTitles:!f,useMargins:!t),query:(language:kuery,query:''),timeRestore:!f,title:'URL%20Drilldown%20Demo',viewMode:view)`)
+----
+
+Instead, each app should expose {kib-repo}tree/{branch}/src/plugins/share/public/url_generators/README.md[a URL generator].
+Other apps should use those URL generators for deep-linking.
+To get a better idea, take a look at {kib-repo}tree/{branch}/src/plugins/dashboard/public/url_generator.ts#L38[dashboard's app URL generator].
+It allows specifying various dashboard app state pieces like: dashboardId, filters, query, time range and more.
+And then others should use it to generate a link a dashboard with specified state:
+
+[source,typescript jsx]
+----
+const dashboardUrl = dashboardUrlGenerator.createUrl({dashboardId, filters, timeRange});
+----
+
+There are two ways to access other's app URL generator in your code:
+
+1. *(preferred)* From a plugin contract of a destination app.
+2. (in case an explicit plugin dependency is not possible) Using {kib-repo}tree/{branch}/src/plugins/share/public/url_generators/README.md[url generator service] from `share` plugin. 
```

Comment:
nit, this link for the second time to the URL generator service, first time is one paragraph above.